crates/full-node/sov-metrics/src/influx_db_nonnative.rs

```diff
@@ -6,58 +8,160 @@ use crate::MaybeTimer;
 use crate::Metric;
 
+type ArcFormatFn =
+    Arc<dyn (Fn(&[u8], &mut fmt::Formatter<'_>) -> fmt::Result) + Send + Sync + 'static>;
+
 /// Metrics for a single state access.
 #[derive(Debug)]
 pub struct StateAccessMetric {
-    #[allow(missing_docs)]
-    pub op: &'static str,
-    #[allow(missing_docs)]
-    pub key_size: usize,
+    /// The key being accessed
+    #[cfg_attr(not(feature = "native"), allow(dead_code))]
+    key: MetricSlotKey,
     #[allow(missing_docs)]
     pub storage_read_size: Option<u32>,
     #[allow(missing_docs)]
     pub duration: MaybeTimer,
+    /// The type of access.
+    pub access_type: StateAccessType,
+}
+
+/// The type of state access.
+#[derive(Debug)]
+pub enum StateAccessType {
+    /// Fetch the size of the value
+    GetSize,
+    /// Fetch the value itself
+    GetValue,
+}
+
+/// A key for a metric.
+pub struct MetricSlotKey {
+    key: Arc<Vec<u8>>,
+    display_fn: Option<ArcFormatFn>,
+}
+
+impl std::fmt::Display for MetricSlotKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(display_fn) = &self.display_fn {
+            display_fn(self.key.as_slice(), f)
+        } else {
+            write!(f, "unknown")
+        }
+    }
+}
+impl std::fmt::Debug for MetricSlotKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(display_fn) = &self.display_fn {
+            write!(f, "MetricKey {{ key: ")?;
+            display_fn(self.key.as_slice(), f)?;
+            write!(f, " }}")
+        } else {
+            write!(f, "MetricKey {{ key: {:?} }}", self.key.as_slice())
+        }
+    }
 }
 
 impl StateAccessMetric {
     /// Creates a new state access metric.
-    pub fn new(op: &'static str, key_size: usize) -> Self {
+    pub fn new_size(key: Arc<Vec<u8>>, display_fn: Option<ArcFormatFn>) -> Self {
         Self {
-            op,
-            key_size,
+            key: MetricSlotKey { key, display_fn },
             storage_read_size: None,
             duration: MaybeTimer::started(),
+            access_type: StateAccessType::GetSize,
+        }
+    }
+
+    /// Creates a new state access metric.
+    pub fn new_read(key: Arc<Vec<u8>>, display_fn: Option<ArcFormatFn>) -> Self {
+        Self {
+            key: MetricSlotKey { key, display_fn },
+            storage_read_size: None,
+            duration: MaybeTimer::started(),
+            access_type: StateAccessType::GetValue,
         }
     }
 
     /// Returns a serializable placeholder metric.
     pub fn placeholder() -> Self {
         Self {
-            op: "placeholder",
-            key_size: 0,
+            key: MetricSlotKey {
+                key: Arc::new(vec![]),
+                display_fn: None,
+            },
             storage_read_size: None,
             duration: MaybeTimer::Completed(Duration::from_secs(0)),
+            access_type: StateAccessType::GetSize,
         }
     }
 }
 #[cfg(feature = "native")]
+#[allow(dead_code)]
```

Comment:
Do we need the dead_code here?

Reply:
Nope, good catch! Leftover from some experimentation